### PR TITLE
Fixture generation framework

### PR DESCRIPTION
**Pull request by peterargue**
This is a proposal for a new model fixtures framework that support generating deterministic test fixtures.

This PR only implements fixture generators required for `execution_data.BlockExecutionData` and `flow.Block` objects. It also includes an implementation of deterministic fixture tests for `BlockExecutionData` as a demonstration.

### CODE REVIEW

File: utils/unittest/fixtures/signer_indices.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// SignerIndicesGenerator generates signer indices with consistent randomness.
+type SignerIndicesGenerator struct {
+}
+
+// Fixture generates signer indices with the specified count.
+// Uses default 10-bit vector size.
+func (g *SignerIndicesGenerator) Fixture(t testing.TB, n int) []byte {
+	indices := bitutils.MakeBitVector(10) // default bit size
+	for range n {
+		bitutils.SetBit(indices, 1)
```
Comment (tarakby) [inline utils/unittest/fixtures/signer_indices.go]: This seems to just set the bit `1` multiple times without randomness. 
I think it's supposed to be something like:
```
for i:= range n {
 if g.Intn(2) == 1 {
    bitutils.SetBit(indices, i)
    }
  }
  ```
  and then the `SignerIndicesGenerator` can embed a `RandomGenerator`

File: utils/unittest/fixtures/signer_indices.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+// SignerIndicesGenerator generates signer indices with consistent randomness.
+type SignerIndicesGenerator struct {
+}
+
+// Fixture generates signer indices with the specified count.
+// Uses default 10-bit vector size.
+func (g *SignerIndicesGenerator) Fixture(t testing.TB, n int) []byte {
+	indices := bitutils.MakeBitVector(10) // default bit size
```
Comment (tarakby) [inline utils/unittest/fixtures/signer_indices.go]:
```suggestion
	indices := bitutils.MakeBitVector(n) // default bit size
```

File: utils/unittest/fixtures/address.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+func ToSDKAddress(addr flow.Address) sdk.Address {
+	var sdkAddr sdk.Address
+	copy(sdkAddr[:], addr[:])
+	return sdkAddr
+}
+
+// CorruptAddress corrupts the first byte of the address.
+func CorruptAddress(t testing.TB, addr flow.Address) flow.Address {
+	addr[0] ^= 1
+	require.False(t, flow.Testnet.Chain().IsValid(addr), "invalid address fixture generated valid address")
```
Comment (tarakby) [inline utils/unittest/fixtures/address.go]: maybe the chain ID can be an input instead of hardcoding Testnet?

File: utils/unittest/fixtures/address.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+	sdk "github.com/onflow/flow-go-sdk"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// AddressGenerator generates addresses with consistent randomness.
+type AddressGenerator struct {
+	randomGen *RandomGenerator
```
Comment (tarakby) [inline utils/unittest/fixtures/address.go]: Would it be better to embed the field so that `AddressGenertor` inherits all the `RandomGenerator` methods? 
This applies to all the generators of the PR. 

```suggestion
	*RandomGenerator
```

File: utils/unittest/fixtures/generators_test.go [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+	suite1 := NewGeneratorSuite(t)
+	suite2 := NewGeneratorSuite(t)
+
+	// generated values should be different
+	header := suite1.BlockHeaders().Fixture(t)
+	header2 := suite2.BlockHeaders().Fixture(t)
+	assert.NotEqual(t, header, header2)
+}
+
+func TestGeneratorsDeterministic(t *testing.T) {
```
Comment (tarakby) [inline utils/unittest/fixtures/generators_test.go]: I believe there is a way to make all these tests into a loop over a slice of interface pointers. This would remove the redundant code and we can extend the tests for future structures easily.

File: utils/unittest/fixtures/random.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// RandomGenerator provides random value generation with consistent randomness.
+type RandomGenerator struct {
+	rng *rand.Rand
```
Comment (tarakby) [inline utils/unittest/fixtures/random.go]: I think this field could be embedded too, and we could remove all explicit definitions below of methods `Uint32`, `Uint64`..

File: utils/unittest/fixtures/generator.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+	}
+}
+
+// RNG returns the random number generator used by this suite.
+func (g *GeneratorSuite) RNG() *rand.Rand {
+	return g.rng
+}
+
+// Random returns a random generator for this suite.
+func (g *GeneratorSuite) Random() *RandomGenerator {
```
Comment (tarakby) [inline utils/unittest/fixtures/generator.go:66]: I got a bit confused with the `Random()` name but not sure there is a better one

File: utils/unittest/fixtures/address.go (hunk 4 above)
Comment (peterargue) [reply utils/unittest/fixtures/address.go]: I think this would reduce some amount of boilerplate, but would also export `RandomGenerator`'s API. I think this could lead to some weird coupling. Since there's no way to embed without exporting the API, my preference is to keep it as a member.

File: utils/unittest/fixtures/address.go (hunk 3 above)
Comment (peterargue) [reply utils/unittest/fixtures/address.go]: good catch.

File: utils/unittest/fixtures/random.go (hunk 6 above)
Comment (peterargue) [reply utils/unittest/fixtures/random.go]: good idea

File: utils/unittest/fixtures/signer_indices.go (hunk 2 above)
Comment (peterargue) [reply utils/unittest/fixtures/signer_indices.go]: this is pulled from the existing test fixture. 
https://github.com/onflow/flow-go/blob/d591c6f885fcde053a749b1ad4438f16087652f1/utils/unittest/fixtures.go#L1078-L1084

happy to change it to whatever is a good valid random value

File: utils/unittest/fixtures/signer_indices.go (hunk 1 above)
Comment (peterargue) [reply utils/unittest/fixtures/signer_indices.go]: similarly, this is from the original fixture.

@jordanschalm any thoughts on how to improve this fixture?

File: utils/unittest/fixtures/README.md [hunk 8]
```diff
@@ -0,0 +1,854 @@
+# Flow Go Fixtures Module
+
+A context-aware test fixture generation system for Flow Go that provides deterministic, reproducible test data with shared randomness across all generators. This module replaces the standalone fixture functions with a comprehensive suite of generator objects.
```
Comment (janezpodhostnik) [inline utils/unittest/fixtures/README.md]: This should probably mention that you should not share a suite across multiple go-routines.

File: utils/unittest/fixtures/generator.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+	t.Logf("generator suite seed: %d", config.seed)
+
+	return &GeneratorSuite{
+		rng: rand.New(rand.NewSource(config.seed)),
+	}
+}
+
+// RNG returns the random number generator used by this suite.
+func (g *GeneratorSuite) RNG() *rand.Rand {
```
Comment (janezpodhostnik) [inline utils/unittest/fixtures/generator.go]: since `*RandomGenerator` returned from `Random` is already a `*rand.Rand` we could just have one method called `RNG` that returns a `*RandomGenerator`

File: utils/unittest/fixtures/identifier.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+// Fixture generates a random identifier.
+func (g *IdentifierGenerator) Fixture(t testing.TB, opts ...func(*identifierConfig)) flow.Identifier {
+	config := &identifierConfig{}
+
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	id, err := flow.ByteSliceToId(g.randomGen.RandomBytes(t, flow.IdentifierLen))
+	require.NoError(t, err)
```
Comment (janezpodhostnik) [inline utils/unittest/fixtures/identifier.go]: I would prefer a simpler generator API when possible. In cases like this where the error can exclusively happen if  there is some serious issue with `ByteSliceToId` I would rather the framework just panicked so that the API does not need to accept the `t testing.TB` parameter. 

I think at least in this case this would be acceptable.

File: utils/unittest/fixtures/address.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+		// we use a 32-bit index - since the linear address generator uses 45 bits,
+		// this won't error
+		index: uint64(g.randomGen.Uint32()),
+	}
+
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	addr, err := config.chainID.Chain().AddressAtIndex(config.index)
```
Comment (janezpodhostnik) [inline utils/unittest/fixtures/address.go:79]: Looks like this also isn't expected to panic since index is `Uint32`.

File: utils/unittest/fixtures/address.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+func ToSDKAddress(addr flow.Address) sdk.Address {
+	var sdkAddr sdk.Address
+	copy(sdkAddr[:], addr[:])
+	return sdkAddr
+}
+
+// CorruptAddress corrupts the first byte of the address.
+func CorruptAddress(t testing.TB, addr flow.Address, chainID flow.ChainID) flow.Address {
+	addr[0] ^= 1
+	require.False(t, chainID.Chain().IsValid(addr), "invalid address fixture generated valid address")
```
Comment (janezpodhostnik) [inline utils/unittest/fixtures/address.go]: Is this possible?

File: utils/unittest/fixtures/block_header.go [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+// WithParentAndSoR returns an option to set the parent and source of randomness of the block header.
+func (g *BlockHeaderGenerator) WithParentAndSoR(parent *flow.Header, source []byte) func(*blockHeaderConfig) {
+	return func(config *blockHeaderConfig) {
+		config.parent = parent
+		config.source = source
+	}
+}
+
+// Fixture generates a basic block header with random values.
+func (g *BlockHeaderGenerator) Fixture(t testing.TB, opts ...func(*blockHeaderConfig)) *flow.Header {
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]: I am wondering if we can omit the parameter `t`. This code hopefully called very often, so lean method signatures might make a difference in practise. In my head, we would _need_ `t` to assess properties of the implementation. However, here, we are using it largely to panic on unexpected problems, which are likely caused by how a test used the Fixture rather than the business logic which the test is validating. Hence, I feel we could just panic inside the fixture. 

nd I feel here, the

File: utils/unittest/fixtures/block_header.go [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	if config.height == 0 && config.view == 0 {
+		// this is the genesis block
+		return g.Genesis(t, opts...)
+	}
+
+	if config.height == 0 || config.view == 0 {
+		t.Fatalf("height and view must either both be greater than 0 or both be 0 (genesis)")
+	}
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]: Conceptually, genesis and spork root blocks are the same. 

I think some of our code might be identifying "genesis" and spork root blocks only by checking the view. I think it is allowed for the height to be bigger than zero. 

I think your fixture is fine ... I am mostly wondering about other engineers  wanting to use your fixture might be tempted to remove this restriction that _both_ must be zero ...what would break if [lines 105-107](https://github.com/onflow/flow-go/blob/46757a3c58e56a227c856643181433b4d23c9038/utils/unittest/fixtures/block_header.go#L105-L107) were removed?

File: utils/unittest/fixtures/block_header.go [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+		},
+		PayloadHash: g.identifierGen.Fixture(t),
+	})
+	require.NoError(t, err)
+	config.parent = parent
+
+	return g.fixtureWithParent(t, config)
+}
+
+// Genesis generates a genesis block header.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]:
```suggestion
// Genesis instantiates a genesis block header. This block has view and height equal to zero. However,
// conceptually spork root blocks are functionally equivalent to genesis blocks. We have decided that
// in the long term, the protocol must support spork root blocks with height _and_ view larger than zero.
```

File: utils/unittest/fixtures/block_header.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+type blockHeaderConfig struct {
+	height  uint64
+	view    uint64
+	chainID flow.ChainID
+	parent  *flow.Header
+	source  []byte
+}
+
+// WithHeight returns an option to set the height of the block header.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]: While I think technically this description is correct, I think the wording could be revised:
* we don't need to say that it "returns an option". We can just say it "is" and option. 
* Furthermore, I think we are going to have a multiple options all called `WithHeight`. This creates ambiguity. When I search for symbols, many IDEs offer the goDoc as part of the search. Hence, it would make sense to say which structure this `WithHeight` is useful for, to make it easier for engineers to choose the applicable option out of the several available. 

```suggestion
// WithHeight is an option for [BlockHeaderGenerator.Fixture]. It sets the height of the block.
```

Analogously for the other `With<..>` options.

File: utils/unittest/fixtures/block_header.go [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+			LastViewTC:         lastViewTC,
+		},
+		PayloadHash: g.identifierGen.Fixture(t),
+	})
+	require.NoError(t, err)
+
+	return header
+}
+
+// List generates a list of block headers.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]: This function is very generic. I would suggest we document the very basic assumptions:
 ```suggestion
// List generates a list of block headers.
//   - All options are applied to each generated block.
//   - The parents of the blocks are individually generated, so blocks don't share a common parent.
//     (unless a parent is explicitly set via an option, in which case all blocks will have this parent).
//   - Views and heights are randomly generated, so there is a small change of views and/or heights
//     being used repeatedly by different blocks.
```
The small change of view and/or height collisions might cause flakey tests.

File: utils/unittest/fixtures/block_header.go [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+	signerIndicesGen *SignerIndicesGenerator
+	quorumCertGen    *QuorumCertificateGenerator
+	timeGen          *TimeGenerator
+}
+
+type blockHeaderConfig struct {
+	height  uint64
+	view    uint64
+	chainID flow.ChainID
+	parent  *flow.Header
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]: I think this might be limiting, because we cannot generate a block with a specified parentID. Instead, the only possibility is to generate the parent first. The current factory pattern for blocks focuses on the values that the child needs to know: https://github.com/onflow/flow-go/blob/6787ab323268c366750bdc7d95d5d65c346bec00/utils/unittest/block.go#L23-L30

Personally, I am in favour of offering to specify a parent header or alternatively just the relevant values. My point is that test logic should not be forced to _always_ specify the parent header explicitly, because lots of tests skip generating the parent and just use an ID instead.

File: utils/unittest/fixtures/block_header.go [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+type BlockHeaderGenerator struct {
+	randomGen        *RandomGenerator
+	identifierGen    *IdentifierGenerator
+	signatureGen     *SignatureGenerator
+	signerIndicesGen *SignerIndicesGenerator
+	quorumCertGen    *QuorumCertificateGenerator
+	timeGen          *TimeGenerator
+}
+
+type blockHeaderConfig struct {
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/block_header.go]: Overall, I have mixed feeling about this approach:
* Take a look at the [blockFactory](https://github.com/onflow/flow-go/blob/6787ab323268c366750bdc7d95d5d65c346bec00/utils/unittest/block.go#L9) for example. There are more setters than the one you implement, such as `WithProposerID` and `WithLastViewTC`
* Overall, I think you should expect that in the long term, we will have a setter for _every_ field of the data structure. Then, would the config be a duplicate of the data structure? Do we need to explicitly replicate the entire data structure? 

The [`blockFactory`](https://github.com/onflow/flow-go/blob/9071c96cbad8d1433060c2aed28c2588e9911655/utils/unittest/block.go#L11) circumvents this problem by using an instance of the target data structure directly as the "config". Could we use the same approach here as well? Alternatively, if you want to retain data during the generation process, that is beyond the fields already included in the target struct, we could do something like the following:

```go
type blockHeaderConfig struct {
	flow.Header

	// additional fields for configuration
	// ⋮
}
```

File: utils/unittest/fixtures/generator.go [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+// Random returns a random generator for this suite.
+func (g *GeneratorSuite) Random() *RandomGenerator {
+	return &RandomGenerator{
+		Rand: g.rng,
+	}
+}
+
+// BlockHeaders returns a block header generator for this suite.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/generator.go]: Confused by the wording "for this suite" and similarly about the other getters' godocs. Is the reference to "this suite" important; otherwise, we could/should omit it for brevity.  Furthermore, for specificity of the documentation, I think we should state the concrete type that the generator produces. 

```suggestion
// BlockHeaders returns this suite's generator for [flow.Header].
```

Analogous suggestion applies to all getters for generators in this file

File: utils/unittest/fixtures/generator.go (hunk 9 above)
Comment (peterargue) [reply utils/unittest/fixtures/generator.go]: I'm fine to remove the extra method. I'd prefer to call it `Random()` so it's clear that it has extra functionality beyond just rng

File: utils/unittest/fixtures/address.go (hunk 12 above)
Comment (peterargue) [reply utils/unittest/fixtures/address.go]: I'm not totally sure. this was taken from the original. I figure it's fine to keep in so we're certain the method is returning invalid data as advertised

File: utils/unittest/fixtures/identifier.go (hunk 10 above)
Comment (peterargue) [reply utils/unittest/fixtures/identifier.go]: that's fair. my only counter argument is that it's convenient for all generators to have the same general prototype, even if the `t` isn't strictly required. I didn't try within this PR, but I can imagine scenarios where using generics with the fixture generators could be useful, but that's really only possible if we maintain a consistent interface.

File: utils/unittest/fixtures/generator.go [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+		transactionSigGen: g.TransactionSignatures(),
+	}
+}
+
+// FullTransactions returns a full transaction generator for this suite.
+func (g *GeneratorSuite) FullTransactions() *FullTransactionGenerator {
+	return &FullTransactionGenerator{
+		TransactionGenerator: g.Transactions(),
+	}
+}
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/generator.go]: I personally find the difference between `Transaction` and `FullTransaction` to be insufficiently explained. This is mostly because a `flow.Transaction` does not represent a transaction in the conceptual sense; rather it represents an ExecutedTransaction and should be called this for clarity. 

in my opinion, `FullTransactions` should be renamed to `ExecutedTransactions` and `FullTransactionGenerator` to `ExecutedTransactionGenerator`

File: utils/unittest/fixtures/README.md (hunk 8 above)
Comment (peterargue) [reply utils/unittest/fixtures/README.md]: I added a couple sections below describing reproducibility and concurrency best practices

File: utils/unittest/fixtures/README.md [hunk 22]
```diff
@@ -0,0 +1,854 @@
...
+Generates `time.Time` values with various options for time-based testing scenarios.
+
+**Options:**
+- `WithBaseTime(baseTime time.Time)`: Sets the base time for generation
+- `WithOffset(offset time.Duration)`: Sets a specific offset from the base time
+- `WithOffsetRandom(max time.Duration)`: Sets a random offset from the base time (0 to max)
+- `WithTimezone(tz *time.Location)`: Sets the timezone for time generation
+
+**Methods:**
+- `Fixture(t testing.TB, opts ...func(*timeConfig))`: Generates a time.Time value with optional configuration
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/README.md:910]:
```suggestion
- `Fixture(t testing.TB, opts ...func(*timeConfig))`: Generates a `time.Time` value with optional configuration
```

File: utils/unittest/fixtures/time.go [hunk 23]
```diff
@@ -0,0 +1,75 @@
+package fixtures
+
+import (
+	"testing"
+	"time"
+)
+
+// TimeGenerator generates time.Time values with consistent randomness.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/time.go]:
```suggestion
// TimeGenerator generates [time.Time] values with consistent randomness.
```

File: utils/unittest/fixtures/time.go [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+
+// WithOffsetRandom returns an option to set a random offset from the base time in the range [0, max).
+func (g *TimeGenerator) WithOffsetRandom(max time.Duration) func(*timeConfig) {
+	return func(config *timeConfig) {
+		offset := time.Duration(g.randomGen.Intn(int(max)))
+		config.offset = offset
+	}
+}
+
+// Fixture generates a time.Time value with optional configuration.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/time.go]:
```suggestion
// Fixture generates a [time.Time] value with optional configuration.
```

File: utils/unittest/fixtures/time.go [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+
+	// Apply timezone if specified
+	if config.timezone != config.baseTime.Location() {
+		config.baseTime = config.baseTime.In(config.timezone)
+	}
+
+	return config.baseTime.Add(config.offset)
+}
+
+// List generates a list of time.Time values.
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/time.go]:
```suggestion
// List generates a list of [time.Time] values.
```

File: utils/unittest/fixtures/README.md [hunk 26]
```diff
@@ -0,0 +1,854 @@
...
+| Full Transactions | `FullTransactions()` | Generate complete transactions |
+| Collections | `Collections()` | Generate collections of transactions |
+| Trie Updates | `TrieUpdates()` | Generate ledger trie updates |
+| Transaction Results | `TransactionResults()` | Generate transaction results |
+| Light Transaction Results | `LightTransactionResults()` | Generate light transaction results |
+| Transaction Signatures | `TransactionSignatures()` | Generate transaction signatures |
+| Proposal Keys | `ProposalKeys()` | Generate proposal keys |
+| Events | `Events()` | Generate events with encoding support |
+| Event Types | `EventTypes()` | Generate event types |
+| Time | `Time()` | Generate time.Time values |
```
Comment (AlexHentschel) [inline utils/unittest/fixtures/README.md]:
```suggestion
| Time | `Time()` | Generate [time.Time] values |
```

File: utils/unittest/fixtures/block_header.go (hunk 16 above)
Comment (peterargue) [reply utils/unittest/fixtures/block_header.go]: since these are methods on the generator struct already, I find specifying the type redundant.

what about something like
```
// WithHeight is an option that sets the height of the block.  
```

File: utils/unittest/fixtures/block_header.go (hunk 17 above)
Comment (peterargue) [reply utils/unittest/fixtures/block_header.go]: I think it probably makes more sense to just update this to produce a chain. I doubt there are many cases where we want a list of random blocks, and it would be a problem if they formed a chain. but I know there are many cases where it is useful to have a chain

File: utils/unittest/fixtures/block_header.go (hunk 14 above)
Comment (peterargue) [reply utils/unittest/fixtures/block_header.go]: I think that's fine

File: utils/unittest/fixtures/generator.go (hunk 21 above)
Comment (peterargue) [reply utils/unittest/fixtures/generator.go]: I'm going to just remove it, and we can add `ExecutedTransactions` later if it's needed

File: utils/unittest/fixtures/identifier.go (hunk 10 above)
Comment (peterargue) [reply utils/unittest/fixtures/identifier.go]: I gave this a shot, and I can just remove `t` from all of the fixtures and use panics instead

File: utils/unittest/fixtures/block_header.go (hunk 19 above)
Comment (peterargue) [reply utils/unittest/fixtures/block_header.go]: that's fair. I think it's fine to use the object itself in most cases.

File: utils/unittest/fixtures/address.go (hunk 12 above)
Comment (tarakby) [reply utils/unittest/fixtures/address.go]: If the input address is valid under that chain, then switching one bit is enough to invalidate it. However the function does not guarantee that the input is valid, so it doesn't hurt to add the extra `require.False` check.

File: utils/unittest/fixtures/signer_indices.go (hunk 2 above)
Comment (AlexHentschel) [reply utils/unittest/fixtures/signer_indices.go]: thanks @tarakby for catching this. I think there is a subtlety with the existing code (already existing prior to peter's generalization): 
* `SignerIndicesFixture(n)` panics for any value larger than 10 
* Essentially the fixture says: I am always generating vectors of size 10 and the caller gets to pick how many of the 10 participants "signed"

Note that:
* we uniformly write bit values `1` for indices 0,...,n-1
* now, if we were to also set the vector length to `n`, we would be always returning vectors entirely filled with one. I don't think that this is what we would want because it's an edge case of the possible value space, which is usually not good to test with. 

   @tarakby that's why I am not convinced by your suggestion ☝️ 

I think a more functionally complete fixture would be:
* the caller specifies the number of authorized signers (currently constant 10) _and_ count of participants that actually signed. Lets denote these quantities `authorizedSigners` and `contributingSigners`.
* I think the fixture _should_ reject inputs with `authorizedSigners < 1` and it _must_ reject inputs where `authorizedSigners` < `contributingSigners`. I think panics with dedicated messages are fine (lets start with this simpler solution vs  adding error returns).
* Note that `authorizedSigners` might be a number that is not an integer multiple of 8. In this case, the returned byte slice must round up.
* I would suggest to improve the fixture to randomly sample the bits that are set to one. I quickly made an implementation using a Fisher-Yates style algorithm for deterministic run time. 


```go

func SignerIndicesFixture2(authorizedSigners, contributingSigners int) []byte {
	if authorizedSigners < 1 {
		panic(fmt.Sprintf("there must be at least one authorized signer, but got %d", authorizedSigners))
	}
	if contributingSigners < 0 {
		panic(fmt.Sprintf("contributingSigners cannot be negative, but got %d", contributingSigners))
	}
	if authorizedSigners < contributingSigners {
		panic(fmt.Sprintf("we cannot have more contributing signers (%d) than there are authorized signers (%d)", contributingSigners, authorizedSigners))
	}

	output := bitutils.MakeBitVector(authorizedSigners) // vector of 0s

	// Randomly flip `contributingSigners` number of bit in output.
	// Note that the slice might be padded with tailing 0s that do not represent authorized signers. This is the case if `authorizedSigners`
	// is not an integer multiple of 8. Hence, we must make sure to not flip those padding bits, i.e. restrict bit flips to the indices in
	// 𝓘 := [0, 2, … authorizedSigners-1]. We employ a Fisher-Yates style algorithm to randomly sample indices without replacement from the
	// interval 𝓘. This algorithm has deterministic time and space complexity O(contributingSigners), and hence avoids the risk of
	// theoretically unbounded runtime of the simpler rejection sampling approach.
	indicesAvailableForFlipping := make([]int, authorizedSigners) // interval 𝓘 := [0, 1, … , authorizedSigners-1]
	for j, _ := range indicesAvailableForFlipping {
		indicesAvailableForFlipping[j] = j
	}

	for k := 0; k < contributingSigners; k++ {
		// For the first loop iteration (k=0) we sample from the entire slice `indicesAvailableForFlipping`.
		// We pick an element at a random position, telling us the index of the bit in `output` to flip.
		// Now, we backfill the sampled position with the last element of the slice.
		l := len(indicesAvailableForFlipping)
		i := rand.Intn(l)                                       // sample index in [0, remainingSampleSpace-1]
		bitutils.SetBit(output, indicesAvailableForFlipping[i]) // flip the bit at the sampled index

		// now, the value at position i has been used, so we backfill it with the last element of the slice
		indicesAvailableForFlipping[i] = indicesAvailableForFlipping[l-1]
		indicesAvailableForFlipping = indicesAvailableForFlipping[:l-1] // shrink the slice by 1
	}
	return output
}

func TestFollowerHappyPath(t *testing.T) {
	t.Run("authorizedSigners smaller 1", func(t *testing.T) {
		assert.Panics(t, func() { SignerIndicesFixture2(0, 0) })
		assert.Panics(t, func() { SignerIndicesFixture2(0, -2) })
		assert.Panics(t, func() { SignerIndicesFixture2(-1, 0) })
		assert.Panics(t, func() { SignerIndicesFixture2(-1, -2) })
	})
	t.Run("too many contributingSigners", func(t *testing.T) {
		assert.Panics(t, func() { SignerIndicesFixture2(7, 8) })
	})
	t.Run("scanning legal input space", func(t *testing.T) {
		// We check values for `authorizedSigners` from 1 to 21
		//
		for authorizedSigners := 1; authorizedSigners <= 21; authorizedSigners++ {
			for contributingSigners := 0; contributingSigners <= authorizedSigners; contributingSigners++ {
				si := SignerIndicesFixture2(authorizedSigners, contributingSigners)
				assert.Equal(t, countSetBitsIn(si), contributingSigners)

			}
		}
	})
}

func countSetBitsIn(byteSlice []byte) int {
	count := 0
	for _, b := range byteSlice {
		for i := 0; i < 8; i++ { // Iterate through each bit of the byte
			if (b>>i)&1 == 1 { // Check if the i-th bit is set
				count++
			}
		}
	}
	return count
}
```

Comment (peterargue) [reply utils/unittest/fixtures/signer_indices.go]: coincidentally, I made a similar change to your suggestion this morning (though less fancy or well documented):

https://github.com/onflow/flow-go/pull/7687/files#diff-dda52784bd938903e9a17a1f2f5f591d27f4e8c9390e38f6895041c5035f4bf1R75-R84

does this my simple version accomplish what's needed?
